src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
     columnName ~ columnType ^^ {
       case(name ~ ty) => Column(name, ty)
     }
+
+  def CategoricalParser = "Categorical" ~> "(" ~> """\d+""".r <~ ")" ^^ { n => MultinomialType(n.toInt) }
+  def BooleanParser = "Boolean" ^^ { s => BooleanType }
+  def dataType = CategoricalParser | BooleanParser
+
   def schemaDeclaration: Parser[SchemaDeclaration] =
-    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ^^ {
-      case (r ~ isQuery ~ "(" ~ attrs ~ ")") => {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
```

Comment:
Yeah, perhaps `::` isn't consistent with the column type declarations.  Let's forget that first idea.